### ISSUE TITLE
Make use of SDPA in TorchText T5 Models

### ISSUE THREAD

**Issue by joecummings** [pytorch/text#2135]

**Comment by yohann-benchetrit** [comment]
[Scaled dot-product attention (SDPA) for T5](https://github.com/pytorch/text/blob/726f5dfcf324dfafe867e3039c2beed993a054a2/torchtext/models/t5/modules.py#L450)  does not currently use memory optimisations, such as [FlashAttention](https://arxiv.org/abs/2205.14135) or [Memory-Efficient Attention](https://arxiv.org/abs/2112.05682).

There is a [memory-efficient implementation of SDPA](https://pytorch.org/docs/master/generated/torch.nn.functional.scaled_dot_product_attention.html?highlight=memory+efficient+attention) in PyTorch, but it uses CUDA fused kernels that do not support arbitrary attention masks yet. Hence, we cannot currently use it to optimise memory usage of T5 attention. Indeed, T5 attention requires Relative Positional Embeddings, typically as an (arbitrary) component of the attention mask (see [Issue 96099](https://github.com/pytorch/pytorch/issues/96099)).

There are [plans](https://github.com/pytorch/pytorch/issues/96099#issuecomment-1458609375) to release a memory-efficient PyTorch SDPA working with arbitrary attention masks. Then, we could use it to enable memory-efficient attention for T5.
